src/routes/file_utils.py

```diff
@@ -0,0 +1,134 @@
+import abc
+import hashlib
+import logging
+import os
+import pathlib
+import shutil
+from io import BytesIO
+
+from typing import Tuple, Optional
+from typing.io import IO
+
+import boto3
+from urllib import request
+
+from boto3_type_annotations.s3.client import Client as S3Client
+from dotenv import load_dotenv
+
+logger = logging.getLogger(__name__)
+load_dotenv()
+
+LOCAL_FILES_DIRECTORY = os.environ.get('LOCAL_FILES_DIRECTORY') or '/tmp/scihive-papers'
+EXTERNAL_BASE_ADDRESS = os.environ.get('EXTERNAL_BASE_ADDRESS') or 'http://localhost:5000'
+
+s3_key = os.environ.get('S3_KEY')
+s3_secret = os.environ.get('S3_SECRET')
+s3_bucket = os.environ.get('S3_BUCKET_NAME')
+
+s3_available = s3_key and s3_secret and s3_bucket
+s3_client: Optional[S3Client] = boto3.client(
+    's3',
+    aws_access_key_id=s3_key,
+    aws_secret_access_key=s3_secret
+) if s3_available else None
+
+
+class FileAccessProvider(metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def exists(self, path: str) -> bool:
+        pass
+
+    @abc.abstractmethod
+    def get_link_to_file(self, path: str) -> str:
+        pass
+
+    @abc.abstractmethod
+    def save_file(self, filename: str, content: IO) -> None:
+        pass
+
+
+class S3FileAccessProvider(FileAccessProvider):
+
+    def __init__(self, s3_client: S3Client, s3_bucket: str, prefix: str) -> None:
+        self._s3_client = s3_client
+        self._s3_bucket = s3_bucket
+        self._prefix = prefix
+
+    def exists(self, path: str) -> bool:
+        prefixed_key = f'{self._prefix}/{path}'
+        response = self._s3_client.list_objects_v2(
+            Bucket=self._s3_bucket,
+            Prefix=prefixed_key,
+        )
+        for obj in response.get('Contents', []):
+            if obj['Key'] == prefixed_key:
+                return True
+        return False
+
+    def get_link_to_file(self, path: str) -> str:
+        return f'https://{self._s3_bucket}.s3.amazonaws.com/{self._prefix}/{path}'
+
+    def save_file(self, filename: str, content: IO) -> None:
+        self._s3_client.upload_fileobj(content, self._s3_bucket, f'{self._prefix}/{filename}')
+
+
+class LocalFileAccessProvider(FileAccessProvider):
+
+    def __init__(self, base_path: str, external_base_address: str) -> None:
+        self._base_path = base_path
+        self._external_base_address = external_base_address
+        # make sure the directory exists
+        pathlib.Path(base_path).mkdir(parents=True, exist_ok=True)
+
+    def exists(self, path: str) -> bool:
+        return os.path.isfile(f'{self._base_path}/{path}')
+
+    def get_link_to_file(self, path: str) -> str:
+        return f'{self._external_base_address}/paper/files/{path}'
```

Comment:
Not very important but it's a bit strange to save the localhost path in the DB.
It's only for local dev, therefore I don't mind keeping it around :)

Reply:
Though in practice, when developing locally this would likely store a path to `localhost:5000`, this isn't coupled with `localhost`. 

The idea is basically this - part of the storage solution involves providing a URL that allows anonymous HTTP access to that resources (`key_to_url`), later used in the FE. We couldn't provide an S3 alternative without somehow solving this as well. The previous solution made sense given that S3 supports anonymous URLs out of the box (not sure if it's a best practice, ideally you would go through the backend anyway to allow access control to stored documents, but this isn't a priority).

Since we're storing these files locally, we need a way to still provide an external URL address. If you were to work with an external deployed server with a different EXTERNAL_BASE_ADDRESS it would also work (though we probably won't do that).

Do you have an alternative idea on how to provide these URLs? Happy to change if something makes more sense.

Reply:
You can store the path in two vars - base source (local, s3 bucket, etc) and relative path. 
I don't think it worth the effort at the moment. We can resolve this comment